options/options.go

```diff
@@ -503,80 +515,282 @@ func (opts *ToolOptions) handleUnknownOption(option string, arg flags.SplitArgum
 	return args, fmt.Errorf(`unknown option "%v"`, option)
 }
 
+// Sets options from the URI. If any options are already set, they are added to the connection string.
+// which is eventually added to the connString field.
+// Most CLI and URI options are normalized in three steps:
+//
+// 1. If both CLI option and URI option are set, throw an erroor if they conflict.
+// 2. If the CLI option is set, but the URI option isn't, set the URI option
+// 3. If the URI option is set, but the CLI option isn't, set the CLI option
+//
+// Some options (e.g. host and port) are more complicated. To check if a CLI option is set,
+// we check that it is not equal to its default value. To check that a URI option is set,
+// some options have an "OptionSet" field.
 func (opts *ToolOptions) setOptionsFromURI(cs connstring.ConnString) error {
 	opts.URI.connString = cs
 
-	// if Connection settings are enabled, then verify that other methods
-	// of specifying weren't used and set timeout
 	if opts.enabledOptions.Connection {
-		switch {
-		case opts.Connection.Host != "":
-			return fmt.Errorf(IncompatibleArgsErrorFormat, "--host")
-		case opts.Connection.Port != "":
-			return fmt.Errorf(IncompatibleArgsErrorFormat, "--port")
-		case opts.Connection.Timeout != 3:
-			return fmt.Errorf(IncompatibleArgsErrorFormat, "--dialTimeout")
-		case opts.Connection.SocketTimeout != 0:
-			return fmt.Errorf(IncompatibleArgsErrorFormat, "--socketTimeout")
-		// TODO: TOOLS-2348 will need to address this new argument
-		case opts.Connection.Compressors != "" && opts.Connection.Compressors != "none":
-			return fmt.Errorf(IncompatibleArgsErrorFormat, "--compressors")
-		}
-		opts.Connection.Timeout = int(cs.ConnectTimeout / time.Millisecond)
-		opts.Connection.SocketTimeout = int(cs.SocketTimeout / time.Millisecond)
+		// Port can be set in --port, --host, or URI
+		// Each host/port pair in the options must match the URI host/port pairs
+		if opts.Port != "" {
```

Comment:
Since this is done, it's done :) But this would have probably been a good place to use codegen from a json/yaml file. It's right on the boundary for size.

Reply:
Good point. I think we can look into that in the ticket for fixing the zero-value problem.